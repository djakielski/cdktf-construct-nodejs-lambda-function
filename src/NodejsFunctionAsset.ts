import * as fs from "fs";
import * as path from "path";
import { AssetType, TerraformAsset } from "cdktf";
import { Construct } from "constructs";
import { BuildOptions, buildSync } from "esbuild";
import { NodejsFunctionProps } from "./index";


export class NodejsFunctionAsset extends Construct {
  public readonly asset: TerraformAsset;
  public readonly bundledPath: string;

  constructor(scope: Construct, id: string, props: NodejsFunctionProps) {
    super(scope, id);


    let workingDirectory: string;
    let distPath: string;
    let entryPoint: string;
    if (fs.lstatSync(props.path).isDirectory()) {
      workingDirectory = path.resolve(props.path);
      entryPoint = "index.ts";
    } else {
      workingDirectory = path.resolve(path.dirname(props.path));
      entryPoint = path.basename(props.path);
    }
    this.bundle(workingDirectory, entryPoint);
    distPath = path.join(workingDirectory, "dist");
    this.bundledPath = distPath;

    this.asset = new TerraformAsset(this, "lambda-asset", {
      path: distPath,
      type: AssetType.ARCHIVE,
    });
  }

  private bundle (workingDirectory: string, entryPoint: string) {
    const options = this.getBuildOptionsFromTsConfigOrDefault(workingDirectory, entryPoint);
    const c = buildSync(options);
    if (c.errors.length > 0) {throw new Error("Compiled with errors");}
    return path.join(workingDirectory, "dist");
  }

  private getBuildOptionsFromTsConfigOrDefault(workingDirectory: string, entryPoint: string) {
    let options: BuildOptions;
    const tsconfig = path.join(workingDirectory, "tsconfig.json");
    if (fs.existsSync(tsconfig)) {
      options = JSON.parse(fs.readFileSync(tsconfig).toString()).compilerOptions;
    } else {
      options = {
        platform: "node",
        minify: true,
        target: "es2020",
        bundle: true,
        format: "cjs",
        sourcemap: "linked",
        outdir: "dist",
      };
    }
    options.absWorkingDir = workingDirectory;
    options.entryPoints = [entryPoint];
    options.outdir = "dist";
    return options;
  }
}