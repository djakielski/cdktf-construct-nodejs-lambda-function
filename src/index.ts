import { DataAwsIamPolicyDocument } from '@cdktf/provider-aws/lib/data-aws-iam-policy-document';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { LambdaFunction, LambdaFunctionConfig } from '@cdktf/provider-aws/lib/lambda-function';
import { Construct } from 'constructs';
import { NodejsFunctionAsset } from './NodejsFunctionAsset';

export interface NodejsFunctionProps {
  readonly path: string;
}

export interface NodeJsLambdaFunctionConfig extends LambdaFunctionConfig {
  readonly sourceCodePath: string;
  readonly nodeVersion?: string; //See identifier on https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html
}

export class NodejsFunction extends LambdaFunction {
  constructor(scope: Construct, id: string, props: NodeJsLambdaFunctionConfig) {
    super(scope, id, props);
    this.handler = 'index.handler';
    this.runtime = props.nodeVersion ?? 'nodejs18.x';
    this.lifecycle = {
      ignoreChanges: ['source_code_hash'],
    };

    const code = new NodejsFunctionAsset(this, id+'-code', { path: props.sourceCodePath });
    this.filename = code.asset.path;
    this.sourceCodeHash = code.asset.assetHash;
    this.role = new IamRole(scope, id+'-role', {
      assumeRolePolicy: new DataAwsIamPolicyDocument(scope, id+'-rolePolicy', {
        statement: [
          {
            sid: 'allowAssume',
            actions: ['sts:AssumeRole'],
            effect: 'Allow',
            principals: [{ type: 'AWS', identifiers: ['lambda.amazonaws.com'] }],
          },
        ],
      }).json,
      name: props.role,
    }).arn;
  }
}