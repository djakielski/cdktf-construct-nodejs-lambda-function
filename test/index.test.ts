import { Testing } from "cdktf";
import {NodejsFunction} from "../src";
import "cdktf/lib/testing/adapters/jest";

expect.addSnapshotSerializer({
    test: (val) => typeof val === 'string',
    print: (val) =>
        `"${(val as string)
            // eslint-disable-next-line no-useless-escape
            .replace(/\"/g, '\\"')
            .replace(/([A-Fa-f0-9]{32})\.(json|zip)/, '[FILENAME REMOVED]')
            .replace(
                /container-assets-\${AWS::AccountId}-\${AWS::Region}:([A-Fa-f0-9]{64})/,
                'container-assets-${AWS::AccountId}-${AWS::Region}:[HASH REMOVED]',
            )
            .replace(
                /lambda-asset_([A-Za-z0-9]{8})\/([A-Za-z0-9]{32})/,
                'lambda-asset_[HASH REMOVED]/[HASH REMOVED]',
            ).replace(
                /([A-Fa-f0-9]{32})/,
                '[HASH REMOVED]',
            )}"`,
});


describe("Test NodejsFunction", () => {
    it("snapshotTest", () => {
        expect(
            Testing.synthScope((scope) => {
                new NodejsFunction(scope, "function", {
                    sourceCodePath: "test-resources/index.ts",
                    functionName: "myNodeJsFunction",
                    role: "myIamRole"
                });
            })
        ).toMatchSnapshot()
    });
});