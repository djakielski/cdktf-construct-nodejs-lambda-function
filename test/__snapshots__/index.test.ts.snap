// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NodejsFunction snapshotTest 1`] = `
"{
  \\"resource\\": {
    \\"aws_lambda_function\\": {
      \\"function\\": {
        \\"filename\\": \\"assets/function_function-code_lambda-asset_[HASH REMOVED]/[HASH REMOVED]/archive.zip\\",
        \\"function_name\\": \\"myNodeJsFunction\\",
        \\"handler\\": \\"index.handler\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"source_code_hash\\"
          ]
        },
        \\"role\\": \\"myIamRole\\",
        \\"runtime\\": \\"nodejs18.x\\",
        \\"source_code_hash\\": \\"[HASH REMOVED]\\"
      }
    }
  }
}"
`;
